Added history interface

Gazing opened the pull request.

Gazing commented:
![sessiontut](https://user-images.githubusercontent.com/9114099/36637552-42eff5be-19ac-11e8-86b8-715ec3b0f433.png)

Ok so I fixed "sessionId" error from popping up. As for the actual logging, I tested it on my local machine and everything works fine, so I'm not sure why its not showing up. Can you maybe check the database and see if anything is created in the collection chatHistory.

hughdani commented:
Not sure what the issue was but clearing what I had in the chatHistory collection fixed it. Would like someone else to double check it just to be sure.

- the previously mentioned console errors still occur if you go to the history page right after creation (no one joins the room)

![image](https://user-images.githubusercontent.com/31714259/36648614-ad3b2968-1a63-11e8-941a-a60624ee2bdf.png)

> - Some alignment issues (recreate by opening the console )
> - We might want to reconsider how we handle message for images
> - The 3rd and 5th message are both latex but in different format. The 3rd is sent from chat and 5th is from math editor
> 

- the students tab table **only** displays messages from System and the admin who is logged in
- search bar under the students tab does not filter correctly (e.g. search for [System] displays an empty table)


Questions: 
- how do I generate more than one session to test "export all sessions"?
- does entering text in the search filter do anything to the export file / should it?

Gazing commented:
Student table filtering now works properly and alignment has been fixed.

I have no idea why students are not showing up for you, but it works fine for me.

Sessions are updated whenever the room settings are changed.
Lastly, right now there is no filtering for the exports but maybe it will be added if its really needed

hughdani commented:
I think the problem is that I am using the link for the instructor which means I do not see posts from the student when accessing the student tab. How do I generate URL for a student?

![image](https://user-images.githubusercontent.com/31714259/36926283-e2b427ea-1e44-11e8-9b1b-18866aea2d20.png)

![image](https://user-images.githubusercontent.com/31714259/36926286-e8aaa4d0-1e44-11e8-8b77-8d6b28103884.png)

Gazing commented:
I added the partial url to the student list you can just use that with the host url and it will let you login as a student. And the sessionid bug should not not be happening. 

As for export all it works fine for me.

hughdani commented:
- The issue with message count seem to be fixed after I entered the url specific for a utorid.
- Export All Sessions seems to work now
- less styles seem to be refactored successfully, but it is weird that files with no changes are still listed
![image](https://user-images.githubusercontent.com/31714259/36951611-c0a56688-1fd3-11e8-9cd6-257287d04f1d.png)

=====================

- (console error) sessionID error no longer appears when clicking the Messages and Student tab but still appears when clicking the Export button

![image](https://user-images.githubusercontent.com/31714259/36951628-f4afb456-1fd3-11e8-98e8-7ea4f056c7cb.png)

- (bug?) in an incognito tab I first entered the url (containing the specific token) for student1, typed something in chat then entered the url (containing the specific token) for student2, without logging out in the same tab. Even when using the url for the second student it still used the the details of student1 when recording history. There was also no prompt for entering a username.

- (formatting) when I open the chat history export for a session, the timestamps item is not represented as a date or time but in a large integer number probably representing seconds since some point in time

- (UX) I find the csv generated from Export All Sessions is difficult to read as everything is appended right after each other. Might not be relevant if its main purpose is data and it can be easily extracted but would likely need a different user's opinion on it.

- (bug) if 2 classrooms with different **Names** are created where both have the same **Room Name**, the history for the second classroom is not recorded. When joining, both are directed to a common url containing the same **Room Name**.

- I am not sure if this is a feature or bug, if a classroom is deleted its history still remains if you recreate the room with the same details

- some related weird behavior, I only see the invalid error for older rooms created without sessionID field as it is null but not for newly created rooms as they is an actual id in the session field. So this feedback error is not usually visible if the database is up to date. Clicking Export All Sessions still generates all the history including old history for that invalid session but clicking Export does not do anything as expected. Might be something to do with the GET console error when viewing that invalid session.

 ![image](https://user-images.githubusercontent.com/31714259/36951955-a394e0c4-1fd7-11e8-8973-8a104aaf7646.png)

![image](https://user-images.githubusercontent.com/31714259/36952097-560e6b98-1fd9-11e8-90f3-b315457c1b87.png)

Gazing commented:
OK, I fixed the sessionId error messages.


Two classrooms shouldn't have the same room name and that has been fixed as well.


In terms of logging in as different students on the same browser, its not possible since its designed to allow them to leave and log in without having to change usernames. You will have to log out first before being able to use another student login.

Formatting wise, the date is left untouched because I don't know what format Sohee might want.

Lastly, I removed the sessions with no sessionId from export all so its more consistent.

hughdani commented:
- (fixed) sessionId console error message doesn't appear anymore for export button
- (fixed) sessions with null sessionId are no longer imported along with valid ones when exporting all

- I was thinking of consistency for the csv date format since the website shows it in human readable date format but we can keep it in seconds in the csv for now and ask her next meeting. We can also ask her about export all formatting.

- I was worried about impersonation issues with logging but it seems like the design works. Is it possible to hide the token so its not in the url? It is important information that can be easily stolen and since that's the main identification we have in case an incident happens, I think it should be more protected.
=============================
- (bug) classroom bug still exists. When creating a classroom with different names but same room name, history only records for the room that was created first. I think this is because the url is the same and  does not take in account of the **name** of the classroom (which is unique) and some conflict happens?
The following console error stops me from creating two rooms with the same **name** but there is no such error for creating two different classrooms with the same **room name**. I am not sure if there is a programmatic difference between name and room name other than room name is used in url construction. 

![image](https://user-images.githubusercontent.com/31714259/36955597-6ef8432c-1ff7-11e8-8121-4a6361542026.png)

- (bug/feature?) convo history is still retained after deleting a classroom and recreating it with the same details for name and room name

- (bug) when using the log out button there is a system message recorded as "null left the room"

Gazing commented:
Weird I thought it already checked for duplicate room names. Well that should be fixed now.
And logout from chat should log the correct message.

Now when deleting a classroom the session history will be marked and will no longer appear in a newly created classroom, but will still be stored in the database

Gazing commented:
Yeah utorid stuff is only set on "new user" event not the "join room" event so it should be an easy fix.

Gazing commented:
Ok, I changed it so that now when a student changes to another classroom, they must login again. This should make it so all student information is consistent when using a different url.

hughdani commented:
Good fix. Minor change left.

- (UX) I get the popup still when I mess up the URL like changing the token.

![image](https://user-images.githubusercontent.com/31714259/37305679-afe15cf6-260b-11e8-9ad1-5dc3917f5ad5.png)

Resolve conflicts to merge.

Gazing commented:
Conflicts resolved.

hughdani commented:
I had to change a few id names regarding the error popups to be specific to login. There was legacy handling in other controllers which used the same style of creating the `<p>` element and appending it to the chat-box div which would have broke. This old stuff will need refactoring in the future.

Feel free to merge :tada:

hughdani commented on AdminView/index.js:
```diff
@@ -541,7 +541,7 @@ AdminView.prototype.setupApi = function () {
         if (isNaN(req.params.id)) return res.status(400).json({status: 400, message: "Invalid id"});
         this.db.collection("chatHistory").findOne({owner: req.session.user.username, className: req.params.name, sessionId: parseInt(req.params.id), deleted: null}, {_id: 0}, function (err, session) {
-            if (err) return res.status(500).json({status: 500, message: "Server error, could not resolve request"});
+            if (err) return res.status(500).json({stexatus: 500, message: "Server error, could not resolve request"});
```
did you mean status?

hughdani commented:
Do you think Sohee might want the TA and instructor flags as part of the export csv for her data?

I will approve when you respond to the possible typo.

Gazing commented:
Fixed the typo and the export already has the the isTA and isInstructor fields, it shows up as 1 if true and empty if its false

hughdani commented:
Since Gulp tools are now active, the location of compiled files have changed. I suggest that you remove the compiled css files and only keep less files. You will need to change css references to point to somewhere in  `public/app/dist/css`. For example previous reference to `public/app/styles/landing/main.css` will need to refer to `public/app/dist/css/styles/landing/main.min.css`
